examples/jsm/tsl/display/FXAANode.js

```diff
@@ -34,279 +34,257 @@ class FXAANode extends TempNode {
 
-		const FxaaPixelShader = Fn( ( [ uv, fxaaQualityRcpFrame, fxaaQualityEdgeThreshold, fxaaQualityinvEdgeThreshold ] ) => {
+		const Sample = Fn( ( [ uv ] ) => {
 
-			const rgbaM = FxaaTexTop( uv ).toVar();
-			const rgbaS = FxaaTexOff( uv, vec2( 0.0, - 1.0 ), fxaaQualityRcpFrame.xy ).toVar();
-			const rgbaE = FxaaTexOff( uv, vec2( 1.0, 0.0 ), fxaaQualityRcpFrame.xy ).toVar();
-			const rgbaN = FxaaTexOff( uv, vec2( 0.0, 1.0 ), fxaaQualityRcpFrame.xy ).toVar();
-			const rgbaW = FxaaTexOff( uv, vec2( - 1.0, 0.0 ), fxaaQualityRcpFrame.xy ).toVar();
-			// . S .
-			// W M E
-			// . N .
+			return textureNode.uv( uv );
 
-			const contrastN = contrast( rgbaM, rgbaN ).toVar();
-			const contrastS = contrast( rgbaM, rgbaS ).toVar();
-			const contrastE = contrast( rgbaM, rgbaE ).toVar();
-			const contrastW = contrast( rgbaM, rgbaW ).toVar();
+		} );
+		
+		const SampleLuminance = Fn( ( [ uv ] ) => {
 
-			const maxValue = max( contrastN, max( contrastS, max( contrastE, contrastW ) ) ).toVar();
+			return dot( Sample( uv ).rgb, vec3( 0.3, 0.59, 0.11 ) );
 
-			// . 0 .
-			// 0 0 0
-			// . 0 .
+		} );
 
-			If( maxValue.lessThan( fxaaQualityEdgeThreshold ), () => {
+		const SampleLuminanceOffset = Fn( ( [ texSize, uv, uOffset, vOffset ] ) => {
 
-				return rgbaM; // assuming define FXAA_DISCARD is always 0
+			const shiftedUv = uv.add( texSize.mul( vec2( uOffset, vOffset ) ) );
+			return SampleLuminance( shiftedUv );
 
-			} );
+		} );
 
-			//
+		const ShouldSkipPixel = Fn( ( [ highest, contrast ] ) => {
 
-			const relativeVContrast = sub( contrastN.add( contrastS ), ( contrastE.add( contrastW ) ) ).toVar();
-			relativeVContrast.mulAssign( fxaaQualityinvEdgeThreshold );
+			const threshold = max( _ContrastThreshold, _RelativeThreshold.mul( highest ) );
+			return contrast.lessThan( threshold );
 
-			// 45 deg edge detection and corners of objects, aka V/H contrast is too similar
+		} );
 
-			If( abs( relativeVContrast ).lessThan( 0.3 ), () => {
+		const DeterminePixelBlendFactor = Fn( ( [ lm, ln, le, ls, lw, lne, lnw, lse, lsw, contrast ] ) => {
 
-				// locate the edge
+			let f = float(2.0).mul( ln.add( le ).add( ls ).add( lw ) );
+			f.addAssign( lne.add( lnw ).add( lse ).add( lsw ) );
+			f.mulAssign( 1.0 / 12.0 );
+			f = abs( f.sub( lm ) );
+			f = clamp( f.div( max (contrast, 0) ), 0.0, 1.0 );
 
-				const x = contrastE.greaterThan( contrastW ).select( 1, - 1 ).toVar();
-				const y = contrastS.greaterThan( contrastN ).select( 1, - 1 ).toVar();
+			const blendFactor = smoothstep( 0.0, 1.0, f );
+			return blendFactor.mul( blendFactor ).mul( _SubpixelBlending );
 
-				const dirToEdge = vec2( x, y ).toVar();
-				// . 2 .      . 1 .
-				// 1 0 2  ~=  0 0 1
-				// . 1 .      . 0 .
+		} );
 
-				// tap 2 pixels and see which ones are "outside" the edge, to
-				// determine if the edge is vertical or horizontal
+		const DetermineEdge = Fn( ( [ texSize, lm, ln, le, ls, lw, lne, lnw, lse, lsw ] ) => {
```

Comment:
You can use just a simple JS function without `Fn()`, Thus you can return an JS object, with any TSL object, preserving the booleans variables. It can replace GLSL structs in this sense.

Reply:
Amazing! Thank you!

Reply:
I guess it makes sense? The JS object will get lost in translation to WGSL/GLSL right?

Reply:
I think you can use a simple JS function like the example below, without `Fn`.

```js
const TestFn = () => {

	return { color: vec4( 1.0, 0.0, 0.0, 1.0 ) };

};

const ApplyFXAA = Fn( ( [ uv, texSize ] ) => {

	return TestFn().color.mul( 0.5 );
	...
```

Reply:
I can't get it to work. I wanted to pass the object between different complex Fn functions, but I can't find a way to do that. I'll just move on. But thanks still. 🙏

Reply:
Could you test again with this approach? https://github.com/mrdoob/three.js/pull/29524/commits/c63942bcd533b5622ac42fc9cf1f96a4a870361e

Reply:
Yees. 🙂